import instance from "@/lib/axios";
import axios from "axios";
import { StoreCustomer, DBResponseCommand } from "@/types";
import { StoreCustomerFormValues } from "@/lib/zod-schemas/storeCustomerSchema";
const API_URL = import.meta.env.VITE_API_STORECUSTOMER_URL;

export async function createStoreCustomerData(values: StoreCustomerFormValues) {
  try {
    const { data } = await instance.post<DBResponseCommand<StoreCustomer>>(
      API_URL,
      values
    );
    return data;
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      // Si el error es de Axios, revisa si hay una respuesta del servidor
      if (error.response) {
        console.error("Backend error:", error.response.data);
        // Retorna un mensaje específico dependiendo del status o el mensaje en data
        const allMessages = Object.values(error.response.data.errors || {})
          .flat()
          .map((msg) => `- ${msg}`)
          .join(", ");
        const message = allMessages || "Unexpected server error";
        throw new Error(message);
      } else if (error.request) {
        // Error en la solicitud pero no hubo respuesta
        console.error("No response received from server");
        throw new Error(
          "No response received from server. Please check your connection."
        );
      } else {
        // Error al configurar la solicitud
        console.error("Request setup error:", error.message);
        throw new Error("Error in request setup: " + error.message);
      }
    } else {
      // Error que no está relacionado con Axios (por ejemplo, errores de JavaScript)
      console.error("General error:", error);
      throw new Error(
        "An unexpected error occurred: " + (error.message || "Unknown error")
      );
    }
  }
}

export async function updateStoreCustomerData(
  id: string,
  values: StoreCustomerFormValues
) {
  try {
    const { data } = await instance.put<DBResponseCommand<StoreCustomer>>(
      `${API_URL}/${id}`,
      values
    );
    return data;
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      // Si el error es de Axios, revisa si hay una respuesta del servidor
      if (error.response) {
        console.error("Backend error:", error.response.data);
        // Retorna un mensaje específico dependiendo del status o el mensaje en data
        const allMessages = Object.values(error.response.data.errors || {})
          .flat()
          .map((msg) => `- ${msg}`)
          .join(", ");
        const message = allMessages || "Unexpected server error";
        throw new Error(message);
      } else if (error.request) {
        // Error en la solicitud pero no hubo respuesta
        console.error("No response received from server");
        throw new Error(
          "No response received from server. Please check your connection."
        );
      } else {
        // Error al configurar la solicitud
        console.error("Request setup error:", error.message);
        throw new Error("Error in request setup: " + error.message);
      }
    } else {
      // Error que no está relacionado con Axios (por ejemplo, errores de JavaScript)
      console.error("General error:", error);
      throw new Error(
        "An unexpected error occurred: " + (error.message || "Unknown error")
      );
    }
  }
}

export async function deleteStoreCustomerData(id: string) {
  try {
    const { data } = await instance.delete<DBResponseCommand<StoreCustomer>>(
      `${API_URL}/${id}`
    );
    return data;
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      // Si el error es de Axios, revisa si hay una respuesta del servidor
      if (error.response) {
        console.error("Backend error:", error.response.data);
        // Retorna un mensaje específico dependiendo del status o el mensaje en data
        const allMessages = Object.values(error.response.data.errors || {})
          .flat()
          .map((msg) => `- ${msg}`)
          .join(", ");
        const message = allMessages || "Unexpected server error";
        throw new Error(message);
      } else if (error.request) {
        // Error en la solicitud pero no hubo respuesta
        console.error("No response received from server");
        throw new Error(
          "No response received from server. Please check your connection."
        );
      } else {
        // Error al configurar la solicitud
        console.error("Request setup error:", error.message);
        throw new Error("Error in request setup: " + error.message);
      }
    } else {
      // Error que no está relacionado con Axios (por ejemplo, errores de JavaScript)
      console.error("General error:", error);
      throw new Error(
        "An unexpected error occurred: " + (error.message || "Unknown error")
      );
    }
  }
}

export async function getStoreCustomerDataById(id: string) {
  try {
    const { data } = await instance.get<StoreCustomer>(`${API_URL}/${id}`);
    return data;
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      // Si el error es de Axios, revisa si hay una respuesta del servidor
      if (error.response) {
        console.error("Backend error:", error.response.data);
        // Retorna un mensaje específico dependiendo del status o el mensaje en data
        const allMessages = Object.values(error.response.data.errors || {})
          .flat()
          .map((msg) => `- ${msg}`)
          .join(", ");
        const message = allMessages || "Unexpected server error";
        throw new Error(message);
      } else if (error.request) {
        // Error en la solicitud pero no hubo respuesta
        console.error("No response received from server");
        throw new Error(
          "No response received from server. Please check your connection."
        );
      } else {
        // Error al configurar la solicitud
        console.error("Request setup error:", error.message);
        throw new Error("Error in request setup: " + error.message);
      }
    } else {
      // Error que no está relacionado con Axios (por ejemplo, errores de JavaScript)
      console.error("General error:", error);
      throw new Error(
        "An unexpected error occurred: " + (error.message || "Unknown error")
      );
    }
  }
}

export async function getAllStoreCustomerData() {
  try {
    const { data } = await instance.get<StoreCustomer[]>(API_URL);
    return data;
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      // Si el error es de Axios, revisa si hay una respuesta del servidor
      if (error.response) {
        console.error("Backend error:", error.response.data);
        // Retorna un mensaje específico dependiendo del status o el mensaje en data
        const allMessages = Object.values(error.response.data.errors || {})
          .flat()
          .map((msg) => `- ${msg}`)
          .join(", ");
        const message = allMessages || "Unexpected server error";
        throw new Error(message);
      } else if (error.request) {
        // Error en la solicitud pero no hubo respuesta
        console.error("No response received from server");
        throw new Error(
          "No response received from server. Please check your connection."
        );
      } else {
        // Error al configurar la solicitud
        console.error("Request setup error:", error.message);
        throw new Error("Error in request setup: " + error.message);
      }
    } else {
      // Error que no está relacionado con Axios (por ejemplo, errores de JavaScript)
      console.error("General error:", error);
      throw new Error(
        "An unexpected error occurred: " + (error.message || "Unknown error")
      );
    }
  }
}
